README.md
# reactjs-pull-refresh

This is a react component for pull down refresh.

## Install

```
npm install reactjs-pull-refresh --save
```

## Example

```
npm install
gulp example
```

http://localhost:9090


## Online Example

http://reactjs-ui.github.io/reactjs-pull-refresh/

## Build Example
第一次需要先执行前两步操作，再执行第三步。以后修改例子后，只需要执行第三步即可

* 创建 gh-pages 分支，**在执行 git subtree add 命令之前，需确保 gh-pages 分支下至少存在一个文件**
```
git branch gh-pages
git checkout gh-pages
rm -rf *     //隐藏文件需要单独删除，结合命令 ls -a
git add -A
git commit -m "clear gh-page"
git push --set-upstream origin gh-pages
vim README.md
//输入一些内容
git add README.md
git commit -m "README.md"
git push
git checkout master
```

* 把分支 gh-pages 添加到本地 subtree 中，执行该命令前，请确保 examples-dist 文件夹不存在

```
git subtree add --prefix=examples-dist origin gh-pages --squash
```
  
* 生成在线 examples
```
gulp example:build
git add -A examples-dist
git commit -m "Update online examples"
git subtree push --prefix=examples-dist origin gh-pages --squash
git push
```

## Usage

```javascript
import React, {Component, PropTypes} from 'react';
import {render} from 'react-dom';
import injectTapEventPlugin from 'react-tap-event-plugin';
import PullRefresh from '../src/scripts/index';
import './sass/example.scss';

// 初始化 tapEvent 事件, 移动端
injectTapEventPlugin();

class PullRefreshSimple extends Component {
  constructor(props) {
    super(props);
    this.state = {
      items: 20,
      hasMore: true
    };
  }

  refreshCallback = () => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        let result = false;
        if (Math.random() > 0.2) {
          result = true;
        }
        if (result) {
          this.setState({
            items: 20,
            hasMore: true
          }, () => {
            resolve();
          });
        } else {
          reject();
        }
      }, 1000);
    }).then(() => {
      console.info('刷新成功！');
    }, (error) => {
      console.info('刷新失败！');
      Promise.error(error);
    });
  };

  loadMoreCallback = () => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        let result = false;
        if (Math.random() > 0.2) {
          result = true;
        }
        if (result) {
          this.setState({
            items: this.state.items + 10,
            hasMore: this.state.items <= 40
          }, () => {
            resolve();
          });
        } else {
          reject();
        }
      }, 1000);
    }).then(() => {
      console.info('加载更多成功！');
    }, (error) => {
      console.info('加载更多失败！');
      Promise.error(error);
    });
  };

  handleTouchTap = (e) => {
    console.info('测试下拉刷新插件是否与 Tap 事件冲突');
  };

  render() {
    let contents = [];
    const {items, hasMore} = this.state;

    for (let i = items; i > 0; i--) {
      if (i < 10) {
        contents.push(<li key={i}><a href="http://www.sina.com">这里放置真实显示的DOM内容</a> {i}</li>);
      } else {
        contents.push(<li key={i} onTouchTap={this.handleTouchTap}>这里放置真实显示的DOM内容 {i}</li>);
      }
    }

    const props = {
      maxAmplitude: 80,
      debounceTime: 30,
      throttleTime: 100,
      deceleration: 0.001,
      refreshCallback: this.refreshCallback,
      loadMoreCallback: this.loadMoreCallback,
      hasMore
    };

    return (
      <PullRefresh {...props}>
        <ol className="example-list">
          {contents.map((item) => {
            return item;
          })}
        </ol>
      </PullRefresh>
    );
  }
}

render(
  <PullRefreshSimple />, document.getElementById('layout')
);

```

## Options

| 选项        | 类型   |  功能  |
| --------   | ----- | ---- |
| className | PropTypes.string| 自定义 className|
| children | PropTypes.node| 待渲染的内容|
| refreshCallback | PropTypes.func| 上拉刷新回调函数，需要是 promise 函数|
| loadMoreCallback | PropTypes.func| 下拉加载更多回调函数，需要是 promise 函数|
| hasMore | PropTypes.bool.isRequired| 是否有更多数据|
| loadMoreThrottle | PropTypes.number| 设置加载更多，距离最底部临界值|
| scrollBar | PropTypes.bool| 是否显示滚动条|
| maxAmplitude | PropTypes.number| 设置上下滑动最大弹性振幅度，单位为像素，默认为 80 像素|
| debounceTime | PropTypes.number| 设置防抖时间|
| throttleTime | PropTypes.number| 设置滑动条移动频率，值越大，移动的越缓慢|
| deceleration | PropTypes.number| 设置弹性滑动持续时间，即滑动停止时，弹性持续的时间，值越大，持续时间越短|
| scrollSpeed   | PropTypes.number | 设置滚动加速度，值越大，滚动越快 |
| thresholdOffset | PropTypes.number | 设置上下移动临界值，移动超过该值，则向上或向下滑动 |
| durationSpeed | PropTypes.number | 滑动持续时间系数，系数越大，持续的时间短 |
| easing  | PropTypes.string | 设置加速方式，默认为匀速，详情查看 https://github.com/component/ease |
| refresh  | PropTypes.bool | 是否显示刷新 |
| loadMore  | PropTypes.bool | 是否加载更多 |

## Build

```
gulp build
```

## Publish

```
npm publish
```

## More React Component

* [reactjs-scroll](https://www.npmjs.com/package/reactjs-scroll) 

## Issue

https://github.com/reactjs-ui/reactjs-pull-refresh/issues

## CHANGELOG

Please view [here](./CHANGELOG.md)


CHANGELOG.md
# React Pull Refresh 更新日志

## version 0.5.1  (2017-9-20)
* 修复 bug #5
* 升级更新 node 依赖包
* 升级到 webpack 3 后修改打包配置
* 修改例子，增加了加载失败触发 Promise.error(error);

## version 0.5.0  (2017-4-26)

* 升级更新 node 依赖包
* 升级到 webpack 2 后修改打包配置
* 升级 eslint 检测代码后，修改部分代码检测规则
* 升级 react 15.5.4 后单独引入 prop-types 

## version 0.4.1  (2016-8-23)

* 增加参数 easing refresh 和 loadMore 来分别设置加速度，是否显示刷新和显示加载更多

## version 0.3.3  (2016-8-9)

* 引入 reactjs-scroll 最新版本
* reactjs-scroll 作为 externals 引入，打包时不再编译 reactjs-scroll

## version 0.3.2  调整代码结构(2016-8-9)

* 引入 reactjs-scroll 最新版本
* reactjs-scroll 作为 externals 引入，打包时不再编译 reactjs-scroll

## version 0.3.0  重新实现下拉刷新 react 组件(2016-8-3)

* 下拉刷新基于 [reactjs-scroll](https://github.com/reactjs-ui/reactjs-scroll) 实现
* 把下拉刷新和加载更多统一到一起实现
* 增加了可配置参数选项
* 完善文档和例子

## version 0.2.0  升级改进(2016-7-19)

* 引入了 [react-hammerjs](https://github.com/JedWatson/react-hammerjs) 来更好的处理事件，以及组件 unmount 时卸载 hammer 实例
* 样式 rc-pull-ptr-icon 和 rc-pull-ptr-loading 改为使用时根据实际再定义
* 优化例子
* 增加了可自定义向下滑动刷新效果例子
* 修复部分 bug
* 完善文档

## version 0.1.0  开启 React Pull Refresh 之旅(2016-7-17)

* React Pull Refresh 0.1.0 发布
* 实现向下滑动重新刷新页面
* 给出基本实现例子
* 完善相关文档



examples/simple.js
import React, {Component} from 'react';
import {render} from 'react-dom';
import injectTapEventPlugin from 'react-tap-event-plugin';
import PullRefresh from '../src/scripts/index';
import './sass/example.scss';

// 初始化 tapEvent 事件, 移动端
injectTapEventPlugin();

class PullRefreshSimple extends Component {
  constructor(props) {
    super(props);
    this.state = {
      items: 2000,
      hasMore: true
    };
  }

  refreshCallback = () => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        let result = false;
        if (Math.random() > 0.2) {
          result = true;
        }
        if (result) {
          this.setState({
            items: 20,
            hasMore: true
          }, () => {
            resolve();
          });
        } else {
          reject(new Error('错误'));
        }
      }, 1000);
    }).then(() => {
      console.info('刷新成功！');
    }, (error) => {
      console.info('刷新失败！');
      Promise.error(error);
    });
  };

  loadMoreCallback = () => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        let result = false;
        if (Math.random() > 0.2) {
          result = true;
        }
        if (result) {
          this.setState({
            items: this.state.items + 10,
            hasMore: this.state.items <= 40
          }, () => {
            resolve();
          });
        } else {
          reject(new Error('错误'));
        }
      }, 1000);
    }).then(() => {
      console.info('加载更多成功！');
    }, (error) => {
      console.info('加载更多失败！');
      Promise.error(error);
    });
  }

  handleTouchTap = (e) => {
    console.info('测试下拉刷新插件是否与 Tap 事件冲突');
  };

  render() {
    const contents = [];
    const {items, hasMore} = this.state;

    for (let i = items; i > 0; i--) {
      if (i < 10) {
        contents.push(<li key={i}><a href="http://www.sina.com">这里放置真实显示的DOM内容</a> {i}</li>);
      } else {
        contents.push(<li key={i} onTouchTap={this.handleTouchTap}>这里放置真实显示的DOM内容 {i}</li>);
      }
    }

    const props = {
      maxAmplitude: 80,
      debounceTime: 30,
      throttleTime: 100,
      deceleration: 0.001,
      refreshCallback: this.refreshCallback,
      loadMoreCallback: this.loadMoreCallback,
      hasMore
    };

    return (
      <PullRefresh {...props}>
        <ol className="example-list">
          {contents.map((item) => {
            return item;
          })}
        </ol>
      </PullRefresh>
    );
  }
}

render(
  <PullRefreshSimple />, document.getElementById('layout')
);
